crates/re_build_tools/src/hashing.rs

```diff
@@ -0,0 +1,159 @@
+        .with_context(|| format!("couldn't open {path:?}"))
+        .unwrap();
+    io::copy(&mut file, &mut hasher)
+        .with_context(|| format!("couldn't copy from {path:?}"))
+        .unwrap();
+
+    encode_hex(hasher.finalize().as_slice())
+}
+
+/// Given a directory path, computes the sha256 hash of the accumulated contents of all of its
+/// files (ordered by filename), and returns an hexadecimal string for it.
+///
+/// This includes files in sub-directories (i.e. it's recursive).
+///
+/// If `extensions` is specified, only files with the right extensions will be hashed.
+/// Specified extensions should include the dot, e.g. `.fbs`.
+pub fn compute_dir_hash<'a>(path: impl AsRef<Path>, extensions: Option<&'a [&'a str]>) -> String {
+    let mut hasher = Sha256::new();
+
+    let path = path.as_ref();
+    for filepath in iter_dir(path, extensions) {
+        let mut file = fs::File::open(&filepath)
+            .with_context(|| format!("couldn't open {filepath:?}"))
+            .unwrap();
+        io::copy(&mut file, &mut hasher)
+            .with_context(|| format!("couldn't copy from {filepath:?}"))
+            .unwrap();
+    }
+
+    encode_hex(hasher.finalize().as_slice())
+}
+
+/// Given a crate name, computes the sha256 hash of its source code (ordered by filename) and
+/// returns an hexadecimal string for it.
+///
+/// This includes the source code of all its direct and indirect dependencies.
+pub fn compute_crate_hash(pkg_name: impl AsRef<str>) -> String {
+    use cargo_metadata::{CargoOpt, MetadataCommand};
+    let metadata = MetadataCommand::new()
+        .features(CargoOpt::AllFeatures)
+        .exec()
+        .unwrap();
+
+    let pkg_name = pkg_name.as_ref();
+    let mut files = Default::default();
+
+    let pkgs = crate::Packages::from_metadata(&metadata);
+    pkgs.track_implicit_dep(pkg_name, &mut files);
+
+    let mut files = files.into_iter().collect::<Vec<_>>();
+    files.sort();
+
+    let hashes = files.into_iter().map(compute_file_hash).collect::<Vec<_>>();
+    let hashes = hashes.iter().map(|s| s.as_str()).collect::<Vec<_>>();
+
+    compute_strings_hash(&hashes)
+}
+
+/// Given a bunch of strings, computes the sha256 hash of their contents (in the order they
+/// were passed in) and returns an hexadecimal string for it.
+pub fn compute_strings_hash(strs: &[&str]) -> String {
+    let mut hasher = Sha256::new();
+
+    for s in strs {
+        hasher.update(s);
+    }
+
+    encode_hex(hasher.finalize().as_slice())
+}
+
+/// Writes the given `hash` at the specified `path`.
+///
+/// Panics on I/O errors.
+///
+/// Use [`read_versioning_hash`] to read it back.
+pub fn write_versioning_hash(path: impl AsRef<Path>, hash: impl AsRef<str>) {
+    let path = path.as_ref();
+    let hash = hash.as_ref();
+
+    let contents = unindent::unindent(&format!(
+        "
+        # This is a sha256 hash for all direct and indirect dependencies of this crate's build script.
```

Comment:
your comment knows waaay more about the usecase than the method name implies :P. Plz tweak either